### ISSUE TITLE
[BUG] Retrained model on new session cause garbage for saving 

### ISSUE THREAD

**Issue by IchiruTake** [tensorflow/tensorflow#53397]
I still get this issues at Tensorflow 2.5+, Tensorflow-GPU 2.5+, Keras-2.4.3. h5py=3.1.0
Even reinstalling Tensorflow, keras, and h5py does not resolve the problem.

The model I made is just a stack of Dense layer without anything special. 
Similar to all people, I can evaluate and predict good in the same training kernel after training. But saving a model by `model.save()` after training, and then reload the trained model for re-training cause weight reinitialization. Despite the arg `compile=True` or `compile=False`, or even `model.reset_states()` or not, The error still occurred

Note 1: This error appeared too long, from 2016 till now but it is not resolved
Note 2: Test have been made on TensorFlow 2.5, 2.6, ... but errors still being found

Similar Issue: # 4875 in Keras (keras-team/keras#4875)

The pipeline is similar. (This model is pure of stack of Dense layers)
1) Create the model
2) Compile
2) Train model
3) Save the model. 

-> The model still fine for prediction

---------------------------------------------------------------------------------

NEW KERNEL:
4) Re-load the model (3)
5) Save the model (4)
-> Still ok

BUT
---------------------------------------------------------------------------------

NEW KERNEL:
4) Re-load the model (3)
5) Retrain the model by `model.fit()` (4) -> Cause new additional weight initializer: 
I believe this is made by weight += new_initialized_weights
6) Save the model --> Model used the weight from beginning at the first step of (5) after retraining

Although I have digged into the source code but no sign of this behaviour `weight += new_initialized_weights` was found. 
But the Keras loss at the first batch during re-training phase is large enough so that I could notify this error. I believe this may come from the global variables state maintained in the source, 

The callbacks contained the same configuration ModelCheckpoint at two phases

**Comment by sushreebarsa** [comment]
@IchiruTake 
In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

**Comment by IchiruTake** [comment]
> @IchiruTake In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

The codebase is relatively large. Attempt basic work on Colab did not represent the behaviour. I am trying to search where does the bug came from and notice you if found.

**Comment by IchiruTake** [comment]
> @IchiruTake In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

Unfortunately, I cannot reproduce the issue: Not by the code but reloading  model after stop and reset the kernel 
Link: https://colab.research.google.com/drive/1aUs3j4aMto9deH6AJBCifhTBFV9gOMS5#scrollTo=RxO3-QcfBmo7

The funny thing that until we reset all kernels and repeat loading the process and retrain model, this issue occured. I have found several issues related and some reported this issue can be found. 

Changing the backend from `tf.keras` to `keras` or `keras` to `tf` does not simply resolve solution. Some suggests, reinstalling or upgrading, but it does not resolve properly, or even re-define the global variables or behavior but no progress made 
Related Issue: keras-team/keras#2378 keras-team/keras#12263

**Comment by IchiruTake** [comment]
> @IchiruTake In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

After cleaning and rebuild all the environments (same TensorFlow, Keras version), I have resolved this issue. Thus, it should be noticed that this issue should be closed. Thank you for your help, I am so appreciated.

**Comment by zaneeto** [comment]
@IchiruTake May I ask if you were able to get this working on google colab? I am encountering the same issue while working on Colab as well as running locally, how did you manage to solve it?

**Comment by IchiruTake** [comment]
Yes. First, the error is made due to the difference process between inference stage and training stage. If error is still available, reinstall and clear all cache would help
You should publish your code to let us know the issue explicitly

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by zaneeto** [comment]
I'm new to this so I don't really know what's wrong, but I'm not using any custom layers or anything like that.
My training code is 
`model = keras.models.Sequential([
    keras.layers.Dense(30000,activation='relu'),
    keras.layers.Dense(100,activation='relu'),
    keras.layers.Dense(1,activation='relu')
])

model.compile(loss='mse',optimizer='Adam')
trainingval = model.fit(X_train, y_train, epochs=30, validation_data=[X_valid,y_valid])`

and then I'm saving it with 
`model.save('my_model')`

Then, I load it with 
`model_new = tf.keras.models.load_model('my_model')`

and then attempt to test a dataset using 
`testing = model_new.evaluate(datatesting_X, datatesting_y)`

I've been working on it in a single file before and it's always been working, but now that I need to save and load it (since split into two files) it's just reinitializing the weights and giving the loss that I would expect for the first epoch. 
I'm working on Google Colab too, so I don't think it's anything to do with the environment. Is there a very obvious solution I'm missing? I've tried most of the solutions I've seen.

**Comment by IchiruTake** [comment]
> I'm new to this so I don't really know what's wrong, but I'm not using any custom layers or anything like that. My training code is `model = keras.models.Sequential([ keras.layers.Dense(30000,activation='relu'), keras.layers.Dense(100,activation='relu'), keras.layers.Dense(1,activation='relu') ])
> 
> model.compile(loss='mse',optimizer='Adam') trainingval = model.fit(X_train, y_train, epochs=30, validation_data=[X_valid,y_valid])`
> 
> and then I'm saving it with `model.save('my_model')`
> 
> Then, I load it with `model_new = tf.keras.models.load_model('my_model')`
> 
> and then attempt to test a dataset using `testing = model_new.evaluate(datatesting_X, datatesting_y)`
> 
> I've been working on it in a single file before and it's always been working, but now that I need to save and load it (since split into two files) it's just reinitializing the weights and giving the loss that I would expect for the first epoch. I'm working on Google Colab too, so I don't think it's anything to do with the environment. Is there a very obvious solution I'm missing? I've tried most of the solutions I've seen.
> 
> EDIT: Tried with RMSProp and SGD optimizers just in case it is an issue with the optimizer, but still doesn't work.

I don't know, if you are facing this issue, re-install TensorFlow. I have tested on TF 2.5 on local and TF 2.7, TF 2.8 on Colab but it still worked. You may want to check whether input pipeline between training and fine-tuning are identical
